### PR TITLE
vulkan: Minor compilation fixes and one possible bug

### PR DESCRIPTION
**Pull request by baryluk**
vulkan: Minor compilation fixes and one possible bug

Do not take address of temporaries.
Use proper casts when casting from Vulkan-Hpp to Vulkan.
Fix use of implicit conversions that make gcc not like it
due to overload ambiguity.
Remove unused bool cast operator, that called non existing
bool operator in other class.

Tested with vulkan 1.1.82 and 1.1.73, and gcc 8.2.

### CODE REVIEW

File: src/libgpu/src/vulkan/vulkan_staging.cpp [hunk 1]
```diff
@@ -28,10 +30,13 @@ Driver::allocTempBuffer(uint32_t size)
    VmaAllocationCreateInfo allocDesc = {};
    allocDesc.usage = VMA_MEMORY_USAGE_CPU_TO_GPU;
 
+
    VkBuffer buffer;
    VmaAllocation allocation;
+   static_assert(sizeof(vk::BufferCreateInfo) == sizeof(VkBufferCreateInfo),
```
Comment (brett19) [inline src/libgpu/src/vulkan/vulkan_staging.cpp]: This is guaranteed by the Vulkan-Hpp spec already.

File: src/libgpu/src/vulkan/vulkan_surface.cpp [hunk 2]
```diff
@@ -181,24 +181,24 @@ Driver::_copySurface(SurfaceObject *dst, SurfaceObject *src, SurfaceSubRange ran
 
    vk::ImageCopy copyRegion;
    if (src->desc.dim == latte::SQ_TEX_DIM::DIM_3D) {
-      copyRegion.srcOffset = { 0, 0, static_cast<int32_t>(range.firstSlice) };
+      copyRegion.srcOffset = VkOffset3D { 0, 0, static_cast<int32_t>(range.firstSlice) };
```
Comment (brett19) [inline src/libgpu/src/vulkan/vulkan_surface.cpp]: Interesting.  This shouldn't be needed I don't think?  Why does it fail?

File: src/libgpu/src/vulkan/vulkan_driver.h [hunk 3]
```diff
@@ -50,11 +51,6 @@ class HashedDesc
       return mHash;
    }
 
-   operator bool() const
```
Comment (brett19) [inline src/libgpu/src/vulkan/vulkan_driver.h]: Why is this removed here?

File: src/libgpu/src/spirv/spirv_cf.cpp [hunk 4]
```diff
@@ -170,18 +170,22 @@ void Transpiler::translateCf_LOOP_START_DX10(const ControlFlowInst &cf)
...
+   }
 
    // If state is set to Inactive, set it to InactiveBreak
-   auto isInactive = mSpv->createBinOp(spv::OpIEqual, mSpv->boolType(), stateVal, mSpv->stateInactive());
-   ifBuilder = &spv::Builder::If { isInactive, spv::SelectionControlMaskNone, *mSpv };
-   mSpv->createStore(mSpv->stateInactiveBreak(), mSpv->stateVar());
-   ifBuilder->makeEndIf();
+   {
+      auto isInactive = mSpv->createBinOp(spv::OpIEqual, mSpv->boolType(), stateVal, mSpv->stateInactive());
+      auto ifBuilder = spv::Builder::If { isInactive, spv::SelectionControlMaskNone, *mSpv };
```
Comment (brett19) [inline src/libgpu/src/spirv/spirv_cf.cpp]: Oops shit.  Nice catch.

File: src/libgpu/src/vulkan/vulkan_driver.h (hunk 3 above)
Comment (baryluk) [reply src/libgpu/src/vulkan/vulkan_driver.h]: It is unused method. What would be even the meaning of bool(hash) ?

File: src/libgpu/src/vulkan/vulkan_surface.cpp (hunk 2 above)
Comment (baryluk) [reply src/libgpu/src/vulkan/vulkan_surface.cpp]:
```
/home/user/decaf-emu/src/libgpu/src/vulkan/vulkan_surface.cpp: In member function ‘vulkan::SurfaceObject* vulkan::Driver::_allocateSurface(const vulkan::SurfaceDesc&)’:
/home/user/decaf-emu/src/libgpu/src/vulkan/vulkan_surface.cpp:480:41: error: ambiguous overload for ‘operator=’ (operand types are ‘vk::Offset3D’ and ‘<brace-enclosed initializer list>’)
    bufferRegion.imageOffset = { 0, 0, 0 };
                                         ^
In file included from /home/user/decaf-emu/src/libgpu/./gpu_vulkandriver.h:4,
                 from /home/user/decaf-emu/src/libgpu/src/vulkan/vulkan_driver.h:5,
                 from /home/user/decaf-emu/src/libgpu/src/vulkan/vulkan_surface.cpp:2:
/usr/include/vulkan/vulkan.hpp:4948:15: note: candidate: ‘vk::Offset3D& vk::Offset3D::operator=(const VkOffset3D&)’
     Offset3D& operator=( VkOffset3D const & rhs )
               ^~~~~~~~
/usr/include/vulkan/vulkan.hpp:4925:10: note: candidate: ‘constexpr vk::Offset3D& vk::Offset3D::operator=(const vk::Offset3D&)’
   struct Offset3D
          ^~~~~~~~
/usr/include/vulkan/vulkan.hpp:4925:10: note: candidate: ‘constexpr vk::Offset3D& vk::Offset3D::operator=(vk::Offset3D&&)’

```

Am I using to old libvulkan?

File: src/libgpu/src/spirv/spirv_cf.cpp (hunk 4 above)
Comment (baryluk) [reply src/libgpu/src/spirv/spirv_cf.cpp]: Yeah, taking an address of the temporary is a bad idea.

File: src/libgpu/src/vulkan/vulkan_staging.cpp (hunk 1 above)
Comment (baryluk) [reply src/libgpu/src/vulkan/vulkan_staging.cpp]: Indeed, you are right. I even see in Vulkan-Hpp


  static_assert( sizeof( BufferCreateInfo ) == sizeof( VkBufferCreateInfo ), "struct and wrapper have different size!" );

So, this is redundant, and all structs are already checked.

File: src/libgpu/src/vulkan/vulkan_driver.h (hunk 3 above)
Comment (brett19) [reply src/libgpu/src/vulkan/vulkan_driver.h]: It used to be used to verify if there is actually a hash stored (the hash uint64_t[2]{0, 0}).  I think I ended up getting rid of it in exchange for checking if the last usage is non-zero.

File: src/libgpu/src/vulkan/vulkan_staging.cpp (hunk 1 above)
Comment (baryluk) [reply src/libgpu/src/vulkan/vulkan_staging.cpp]: Removed.

File: src/libgpu/src/vulkan/vulkan_surface.cpp (hunk 2 above)
Comment (brett19) [reply src/libgpu/src/vulkan/vulkan_surface.cpp]: I'm using the latest SDK (effectively, I custom built it for transform feedback).  That's a stupid warning though, since the ambiguous overload it mentions are for the same thing....  They both initialize the same vk type >_<.

**Comment by brett19** [comment]
Could you squash the commits together (as we only do rebase merges) and append the `vulkan:` prefix to the commit message to fit with the commit message styling already in use by the project?

File: src/libgpu/src/vulkan/vulkan_surface.cpp (hunk 2 above)
Comment (brett19) [reply src/libgpu/src/vulkan/vulkan_surface.cpp]: In hindsight, I don't think the extra specificity is actually a bad thing.  You will note that in almost every case, I try to avoid using initializer lists anyways, since the parameters being set are so unclear when you read.

Comment (baryluk) [reply src/libgpu/src/vulkan/vulkan_surface.cpp]: I checked my copy in of Offset3D constructors and the one in master branch of Vulkan-Hpp, and they look identical:

    Offset3D( int32_t x_ = 0,
              int32_t y_ = 0,
              int32_t z_ = 0 );
    explicit Offset3D( Offset2D const& offset2D,
                       int32_t z_ = 0 );
    Offset3D( VkOffset3D const & rhs );

and operator = is also the same:

    Offset3D& operator=( VkOffset3D const & rhs );


so I think it is not related to version of vulkan.

The problem is that the compiler do not know which version to call operator=( VkOffset3D const & rhs ); or compiler provided operator= methods.

This is because due to implicit argument type conversion. the brace-enclosed init list can both be passed to constructors of VkOffset3D and vk::Offset3D.  (this is because VkOffset3D has trivial constructor that can do it).

Yes, the bodies and semantic are the exactly same, but compiler doesn't know that. So you just need to select one you prefer.

I can make it vk::Offset3D { ... }.  I guess it might be more consistent with rest of the code.

I just tested it and it also work.

Comment (brett19) [reply src/libgpu/src/vulkan/vulkan_surface.cpp]: Yup.  Use vk::Offset3D and it will be good.  Then just the commit squash / message and I can merge once CI finishes :)

**Comment by baryluk** [comment]
squashed and edited the commit message.

I do not like doing `git push -f`, but `git rebase -i` + squash looks awesome. Never tried it before I think.

**Comment by brett19** [comment]
Yea, usually you would avoid doing a force push, but in this case, since its a PR specifically, it makes a lot of sense to clean it up before we push it to the 'canonical' repository :)

File: src/libgpu/src/vulkan/vulkan_surface.cpp (hunk 2 above)
Comment (brett19) [inline src/libgpu/src/vulkan/vulkan_surface.cpp]: here

File: src/libgpu/src/vulkan/vulkan_surface.cpp [hunk 5]
```diff
@@ -181,24 +181,24 @@ Driver::_copySurface(SurfaceObject *dst, SurfaceObject *src, SurfaceSubRange ran
...
       copyRegion.srcSubresource = { copyAspect, 0, 0, 1 };
       // range.numSlices is expressed by the extent here...
    } else {
-      copyRegion.srcOffset = { 0, 0, 0 };
+      copyRegion.srcOffset = VkOffset3D { 0, 0, 0 };
```
Comment (brett19) [inline src/libgpu/src/vulkan/vulkan_surface.cpp]: here

File: src/libgpu/src/vulkan/vulkan_surface.cpp [hunk 6]
```diff
@@ -181,24 +181,24 @@ Driver::_copySurface(SurfaceObject *dst, SurfaceObject *src, SurfaceSubRange ran
...
       copyRegion.srcSubresource = { copyAspect, 0, range.firstSlice, range.numSlices };
    }
    if (dst->desc.dim == latte::SQ_TEX_DIM::DIM_3D) {
-      copyRegion.dstOffset = { 0, 0, static_cast<int32_t>(range.firstSlice) };
+      copyRegion.dstOffset = VkOffset3D { 0, 0, static_cast<int32_t>(range.firstSlice) };
```
Comment (brett19) [inline src/libgpu/src/vulkan/vulkan_surface.cpp]: here

File: src/libgpu/src/vulkan/vulkan_surface.cpp [hunk 7]
```diff
@@ -181,24 +181,24 @@ Driver::_copySurface(SurfaceObject *dst, SurfaceObject *src, SurfaceSubRange ran
...
       copyRegion.dstSubresource = { copyAspect, 0, 0, 1 };
       // range.numSlices is expressed by the extent here...
    } else {
-      copyRegion.dstOffset = { 0, 0, 0 };
+      copyRegion.dstOffset = VkOffset3D { 0, 0, 0 };
```
Comment (brett19) [inline src/libgpu/src/vulkan/vulkan_surface.cpp]: and here

File: src/libgpu/src/vulkan/vulkan_surface.cpp (hunk 2 above)
Comment (baryluk) [reply src/libgpu/src/vulkan/vulkan_surface.cpp]: Done

**Comment by brett19** [comment]
Didn't bother waiting for CI, since CI doesn't check the Vulkan stuff yet anyways.
